client/state/index.js

```diff
@@ -132,5 +143,75 @@ export function createReduxStore( initialState = {} ) {
 		isBrowser && window.devToolsExtension && window.devToolsExtension()
 	].filter( Boolean );
 
-	return compose( ...enhancers )( createStore )( reducer, initialState );
+	return enhancers;
 }
+
+/**
+ * Named export function for creating a redux store.
+ * This creates a new redux store within the calypsoStore instance.
+ *
+ * @param { object } initialState Redux state with which to initialize (optional).
+ * @param { array } startingReducers Initial list of reducers to start with.
+ * @param { array } enhancers Assembled list of enhancers for the store.
+ * @return { object } The redux store that was created.
+ */
+export function createReduxStore( initialState, startingReducers = initialReducers, enhancers = createEnhancers() ) {
+	// This creates a normal redux store, but it adds our list of original reducers.
+	// This is so we can modify that list of reducers later on and regenerate it.
+	const reduxStore = compose( ...enhancers )( createStore )( combineReducers( startingReducers ), initialState );
+	reduxStore.reducers = startingReducers;
+
+	/**
+	 * Adds a reducer dynamically to the store.
+	 *
+	 * @param { string } name The top-level context name for the reducer.
+	 * @param { function } func The reducer function, in the format of ( state, action ), returning state.
+	 */
+	function addReducer( name, func ) {
+		if ( reduxStore.reducers[ name ] !== undefined ) {
+			throw new Error( `addReducer(): Reducer by name "${ name }" already in use` );
+		}
+		const newReducerList = { ...reduxStore.reducers, [ name ]: func };
+
+		reduxStore.replaceReducer( combineReducers( newReducerList ) );
+		reduxStore.reducers = newReducerList;
+	}
+
+	/**
+	 * Removes a reducer dynamically from the store.
+	 *
+	 * @param { string } name The top-level context name for the reducer.
+	 */
+	function removeReducer( name ) {
+		const { [ name ]: removedReducer, ...remainingReducers } = reduxStore.reducers;
+
+		if ( ! removedReducer ) {
+			debug( `removeReducer(): Reducer not found by name "${ name }".` );
+			return;
+		}
+
+		reduxStore.replaceReducer( combineReducers( remainingReducers ) );
+		reduxStore.reducers = remainingReducers;
+	}
+
+	/**
+	 * Gets the current list of reducers.
+	 *
+	 * @return { object } The mapping of top-level reducers that are running within the store.
+	 */
+	function getReducers() {
+		return reduxStore.reducers;
+	}
+
+	// Here we return a wrapped redux store.
+	// We omit `replaceReducer` and add `addReducer`, `removeReducer` and `getReducers`.
+	const { replaceReducer, reducers, ...reduxStoreInterface } = reduxStore; // eslint-disable-line no-unused-vars
+
+	return {
+		...reduxStoreInterface,
+		addReducer,
+		removeReducer,
+		getReducers,
+	};
+}
```

Comment:
> That's why I added the list of reducers to the store itself. Otherwise, the responsibility of keeping them together just gets pushed down the line to the calling code, and that's what I feel we should avoid.

sure - that works! the point is getting them out as open as possible - with as few actual dependencies as is possible.

> having no global state

but we're still dealing with considerable shared mutable state when we start manipulating `this` - if we can return entirely new objects when mutating (preserving the old ones) then we can free ourselves from that shared state entirely